### DISCUSSION TITLE
Is there any way to detect On-Demand Revalidation in getStaticProps?

### DISCUSSION THREAD

**Discussion by cjkihl** [vercel/next.js#42673] [Help]
Just like getStaticProps have a property "preview" to detect preview-mode,
Is there any way to detect an on-demand revalidation in getStaticProps? 

When we are building our site, (Around 100.000 pages) we're having a cache for our common data that is used on all pages so we don't have to fetch it over and over in getStaticProps.

However, on revalidation, we want to make sure everything comes from the API.

For example:

```
export const getStaticProps = ({ preview, revalidate }) => {
  if(preview || revalidate) {
    // fetch from API
  } else {
     // fetch from cache
  }
}
```
Is this possible in some way?

**Comment by icyJoseph** [comment]
During the build process there's an ENV var called `NEXT_PHASE`:

```js
console.log({ phase: process.env.NEXT_PHASE });
```

You could read it inside `getStaticProps` to determine is you are in build mode. At runtime this variable is not present, so if the value of `NEXT_PHASE` is anything but `phase-production-build`, then you can assume it is happening at runtime (or during dev mode). You don't need to know the exact string:

```js
import { PHASE_PRODUCTION_BUILD } from "next/constants";

// later on
export const getStaticProps = ({ preview, revalidate }) => {
 const isBuildStep = process.env.NEXT_PHASE === PHASE_PRODUCTION_BUILD
```

Unfortunately the NEXT_PHASE env var is `undefined` on `next dev` and `next start`, you can change that by using `next.config.js` in function mode:

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  experimental: { appDir: true, allowMiddlewareResponseBody: true },
};

module.exports = (phase) => {
  process.env.NEXT_CONFIG_PHASE = phase;
  return nextConfig;
};
```

And now you can compare `process.env.NEXT_CONFIG_PHASE` to `PHASE_PRODUCTION_SERVER` from `next/constants`.

There's perhaps some gotchas, and maybe the fact that the functions that revalidate content run in separate tasks when hosted on VERCEL, might break the `next.config.js` approach.  If that's the case you need to read a bit into how to add env vars via the Webpack object in `next.config`.  Something like:

```js
module.exports = {
  webpack: (config, { buildId, dev, isServer, defaultLoaders, webpack }) => {
    const customPlugins = [
      new webpack.DefinePlugin({
        'process.env.YOUR_BUILD_ID_VAR_NAME': JSON.stringify(buildId),
      })
    ];
    config.plugins.push(...customPlugins);
    // Important: return the modified config
    return config
  },
}
```
